agent/provider/src/payments/payments.rs

```diff
@@ -579,12 +679,22 @@ impl Handler<UpdateCost> for Payments {
                     } else {
                         myself.agreements
                             .get_mut(&msg.invoice_info.agreement_id)
-                            .map(|agreement| agreement.last_send_debit_note = Utc::now());
+                            .map(|agreement| agreement.last_send_debit_note = last_debit_note + update_interval);
                     }
 
                     // Don't bother, if previous debit note was sent successfully or not.
                     // Schedule UpdateCost for later.
-                    ctx.notify_later(msg, update_interval);
+
+                    let delta = Utc::now() - started;
+                    let delay = if update_interval > delta {
+                        // schedule the next debit note as close to `update_interval` as possible
+                        (update_interval - delta).to_std()?
+                    } else {
+                        // the process lasted longer than `update_interval`, continue immediately
+                        Duration::from_secs(0)
+                    };
+
+                    ctx.notify_later(msg, delay);
```

Comment:
It seems for me, that if `ctx.notify_later(msg, delay);` won't be exact enough, the error will accumulate, because we don't depend on value stored in `agreement.last_send_debit_note`. 

Am I right?

Reply:
Is it going to surpass the average network delay though? Assuming the previous solution was good enough, this one is bit more accurate. Using `agreement.last_send_debit_note` is more complex - e.g. when the `ctx` is blocked for time `t` we need to calculate the n-th next timestamp depending on `t`.

This piece of code could have been refactored into `actix::utils::IntervalFunc`; cancellation is a bit more tricky there but we would be free from tracking and calculating timestamps. I didn't want to do that in this PR, could be done in a follow up if you wish.

Reply:
Previous code was good enough, because we didn't track these intervals in relation to Agreement creation timestamp. I'm wondering how big error could accumulate after month of work for example.

My concern is: Is it possible that Agreement will be broken by someone due to this error? If not, we can ignore this problem

Reply:
It’s tightly connected with the debit note acceptance timeout. Previously it was equally likely for that timestamp to drift off.

Let’s fix it in a follow-up PR

Reply:
Ok, please make issue for this